<a href="https://colab.research.google.com/github/kixtter/AnalisisInteligenteDatos/blob/main/Unidad%201/Caso_10_Regresi%C3%B3n_logistica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Caso 10. Regresión Logística
Yander Alexis Avila Medina

##Objetivo
Construir e interpretar modelos de regresión logística

##Descripción
Cargar librerías

Cargar datos

Datos de entrenamiento y validación 

Crear el modelo

Predecir con el modelo

Evaluar el modelo conforme a caractterísticas de MATRIZ DE CONFUSION

##Marco conceptual
El set de datos [Spam](https://github.com/JoaquinAmatRodrigo/Estadistica-machine-learning-python/tree/master/data), obtenido de [UCI Repository Of Machine Learning Databases](http://www.ics.uci.edu/~mlearn/MLRepository.html) contiene información sobre 4601 correos electrónicos clasificados como spam y no spam.

Para cada correo electrónico se dispone de 58 variables: las 48 primeras contienen la frecuencia con la que aparecen en el texto del email determinadas palabras. Las variables 49-54 indican la frecuencia con la que aparecen los caracteres ;’, ‘(’, ‘[’, ‘!’, ‘$’, ‘#’. Las variables 55-57 contienen la media, a longitud máxima y el número total de letras mayúsculas.

Se crear un modelo de regresión logística con el objetivo de clasificar si un correo es spam o no. 

##Librerias

In [4]:
# Tratamiento de datos
import pandas as pd
import numpy as np

# Gráficos
import matplotlib.pyplot as plt
from matplotlib import style
import seaborn as sns

# Preprocesado y modelado
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import plot_confusion_matrix
import statsmodels.api as sm
import statsmodels.formula.api as smf

# Configuración matplotlib
plt.rcParams['image.cmap'] = "bwr"
plt.rcParams['savefig.bbox'] = "tight"
style.use('ggplot') or plt.style.use('ggplot')


##Datos

In [7]:
datos = pd.read_csv('https://raw.githubusercontent.com/JoaquinAmatRodrigo/' \
       + 'Estadistica-machine-learning-python/master/data/spam.csv')
datos['type'] = np.where(datos['type'] == 'spam', 1, 0)

print('Porcentaje de correo spam y no spam')
print(100 * datos['type'].value_counts(normalize = True))

Porcentaje de correo spam y no spam
0    60.595523
1    39.404477
Name: type, dtype: float64


Con esto podemos ver como el 60.60% de los correos no son spam y que el 39.40% si lo son.

##Datos de entrenamiento y validación

In [13]:
X = datos.drop(columns = 'type')
Y = datos['type']

X_train, X_test, y_train, y_test = train_test_split(
    X,
    Y.values.reshape(-1,1),
    train_size = 0.8,
    random_state = 1234, 
    shuffle = True)

##Modelo de regresión logística

In [15]:
X_train = sm.add_constant(X_train, prepend=True)
modelo = sm.Logit(endog = y_train, exog = X_train,)
modelo = modelo.fit()

print(modelo.summary())

Optimization terminated successfully.
         Current function value: 0.200751
         Iterations 15
                           Logit Regression Results                           
Dep. Variable:                      y   No. Observations:                 3680
Model:                          Logit   Df Residuals:                     3622
Method:                           MLE   Df Model:                           57
Date:                Sun, 07 Nov 2021   Pseudo R-squ.:                  0.7009
Time:                        05:09:04   Log-Likelihood:                -738.76
converged:                       True   LL-Null:                       -2469.6
Covariance Type:            nonrobust   LLR p-value:                     0.000
                        coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5530      0.156     -9.956      0.000      -1.859      -1.247
make   

##Predicciones


In [18]:
# Predicciones con intervalo de confianza 
predicciones = modelo.predict(exog = X_train)

# Clasificación predicha
clasificacion = np.where(predicciones<0.5, 0, 1)
clasificacion

array([0, 0, 0, ..., 1, 1, 0])

##Precisión de la prueba

In [19]:
# Precisión de test del modelo 
X_test = sm.add_constant(X_test, prepend=True)
predicciones = modelo.predict(exog = X_test)
clasificacion = np.where(predicciones<0.5, 0, 1)
precision = accuracy_score(
            y_true    = y_test,
            y_pred    = clasificacion,
            normalize = True
           )
print("")
print(f"La precisión de test es: {100 * precision}%")


La precisión de test es: 92.83387622149837%


##Matriz de confusión

In [20]:
# Matriz de confusión de las predicciones de test
confusion_matrix = pd.crosstab(
    y_test.ravel(),
    clasificacion,
    rownames=['Real'],
    colnames=['Predicción']
)
confusion_matrix

Predicción    0    1
Real                
0           535   28
1            38  320

#Conclusiones

El modelo de regresión logística en esta caso nos ayudo a poder determinar cuando es un correo normal y cuando es un correo spam lo cual esto nos permite poder ignorar correos que no son de nuestro interes o tener cuidado ya que pueden ser estafas.

El porcentaje de certeza fue del 92.8%, un valor bastante alto y con esto nos da la seguridad que se si le damos datos para que realice su tratamiento poder dar con aquellos que si son spam sin problema aunque tambien que solo unos pocos se nos escapen o caigan dentro de los que nosotros podemos considerar como correctos.